File: algo/src/test/java/com/hartwig/actin/algo/evaluation/composite/WarnIfTest.java
package com.hartwig.actin.algo.evaluation.composite;

import static com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation;

import static org.junit.Assert.assertTrue;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.TestDataFactory;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.datamodel.ImmutableEvaluation;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.algo.evaluation.TestEvaluationFunctionFactory;

import org.junit.Test;

public class WarnIfTest {

    @Test
    public void canWarnIf() {
        PatientRecord patient = TestDataFactory.createProperTestPatientRecord();

        assertEvaluation(EvaluationResult.WARN, new WarnIf(TestEvaluationFunctionFactory.pass()).evaluate(patient));
        assertEvaluation(EvaluationResult.WARN, new WarnIf(TestEvaluationFunctionFactory.warn()).evaluate(patient));
        assertEvaluation(EvaluationResult.PASS, new WarnIf(TestEvaluationFunctionFactory.fail()).evaluate(patient));
        assertEvaluation(EvaluationResult.UNDETERMINED, new WarnIf(TestEvaluationFunctionFactory.undetermined()).evaluate(patient));
        assertEvaluation(EvaluationResult.NOT_IMPLEMENTED,
                new WarnIf(TestEvaluationFunctionFactory.notImplemented()).evaluate(patient));
        assertEvaluation(EvaluationResult.NOT_EVALUATED, new WarnIf(TestEvaluationFunctionFactory.notEvaluated()).evaluate(patient));
    }

    @Test
    public void canFlipMessagesOnFail() {
        EvaluationFunction fail = record -> ImmutableEvaluation.builder()
                .result(EvaluationResult.FAIL)
                .addFailSpecificMessages("fail 1")
                .addUndeterminedSpecificMessages("undetermined 1")
                .addPassSpecificMessages("pass 1")
                .build();

        Evaluation result = new WarnIf(fail).evaluate(TestDataFactory.createMinimalTestPatientRecord());

        assertTrue(result.passSpecificMessages().contains("fail 1"));
        assertTrue(result.failSpecificMessages().isEmpty());
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/composite/WarnIfTest.kt
package com.hartwig.actin.algo.evaluation.composite

import com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation
import com.hartwig.actin.algo.evaluation.TestEvaluationFunctionFactory
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.TestPatientFactory
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.algo.EvaluationResult
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class WarnIfTest {

    @Test
    fun canWarnIf() {
        val patient: PatientRecord = TestPatientFactory.createProperTestPatientRecord()
        assertEvaluation(EvaluationResult.WARN, WarnIf(TestEvaluationFunctionFactory.pass()).evaluate(patient))
        assertEvaluation(EvaluationResult.WARN, WarnIf(TestEvaluationFunctionFactory.warn()).evaluate(patient))
        assertEvaluation(EvaluationResult.PASS, WarnIf(TestEvaluationFunctionFactory.fail()).evaluate(patient))
        assertEvaluation(EvaluationResult.PASS, WarnIf(TestEvaluationFunctionFactory.undetermined()).evaluate(patient))
        assertEvaluation(EvaluationResult.PASS, WarnIf(TestEvaluationFunctionFactory.notEvaluated()).evaluate(patient))
    }

    @Test
    fun canMoveMessagesToWarnOnPass() {
        val result: Evaluation =
            WarnIf(TestEvaluationFunctionFactory.pass()).evaluate(TestPatientFactory.createMinimalTestWGSPatientRecord())
        assertThat(result.passSpecificMessages).isEmpty()
        assertThat(result.passGeneralMessages).isEmpty()
        assertThat(result.warnSpecificMessages).isNotEmpty()
        assertThat(result.warnGeneralMessages).isNotEmpty()
    }

    @Test
    fun `Should not return inclusion or exclusion molecular events`(){
        val result: Evaluation =
            WarnIf(TestEvaluationFunctionFactory.pass()).evaluate(TestPatientFactory.createMinimalTestWGSPatientRecord())
        assertThat(result.inclusionMolecularEvents).isEmpty()
        assertThat(result.exclusionMolecularEvents).isEmpty()
    }
}